src/storable.rs

```diff
@@ -568,40 +571,192 @@ pub(crate) const fn bounds<A: Storable>() -> Bounds {
+        _ => u32::from_be_bytes([src[0], src[1], src[2], src[3]]) as usize,
     }
 }
 
-pub(crate) const fn bytes_to_store_size(bounds: &Bounds) -> u32 {
+/// Encodes `size` at the beginning of `dst` of length `bytes_to_store_size` bytes.
+fn encode_size(dst: &mut [u8], size: usize, bytes_to_store_size: usize) {
+    match bytes_to_store_size {
+        1 => dst[0] = size as u8,
+        2 => dst[0..2].copy_from_slice(&(size as u16).to_be_bytes()),
+        _ => dst[0..4].copy_from_slice(&(size as u32).to_be_bytes()),
+    };
+}
+
+pub(crate) const fn bytes_to_store_size_bounded(bounds: &Bounds) -> u32 {
     if bounds.is_fixed_size {
         0
-    } else if bounds.max_size <= u8::MAX as u32 {
+    } else {
+        bytes_to_store_size(bounds.max_size as usize) as u32
+    }
+}
+
+const fn bytes_to_store_size(bytes_size: usize) -> usize {
+    if bytes_size <= u8::MAX as usize {
         1
-    } else if bounds.max_size <= u16::MAX as u32 {
+    } else if bytes_size <= u16::MAX as usize {
         2
     } else {
         4
     }
 }
+
+fn encode_size_lengths(sizes: Vec<usize>) -> u8 {
+    assert!(sizes.len() <= 4);
+
+    let mut size_lengths_byte: u8 = 0;
+
+    for size in sizes.iter() {
+        let size_length = bytes_to_store_size(*size);
+        // Number of bytes required to store the size of every
+        // element is represented with 2 bits.
+        size_lengths_byte <<= 2;
+        // `size_length` can take value in {1, 2, 4}, but to
+        // compress it into 2 bit we will decrement its value.
+        size_lengths_byte += (size_length - 1) as u8;
+    }
+
+    size_lengths_byte
+}
+
+fn decode_size_lengths(mut encoded_bytes_to_store: u8, number_of_encoded_lengths: u8) -> Vec<u8> {
+    assert!(number_of_encoded_lengths <= 4);
+
+    let mut bytes_to_store_sizes = vec![];
+
+    for _ in 0..number_of_encoded_lengths {
+        // The number of bytes required to store the size of every
+        // element is represented with 2 bits. Hence we use
+        // mask `11`, equivalent to 3 in the decimal system.
+        let mask: u8 = 3;
+        // The number of bytes required to store size can take value
+        // in {1, 2, 4}, but to compress it to 2-bit,
+        // when encoding we decreased the value, hence now we need
+        // to do inverse.
+        let bytes_to_store: u8 = (encoded_bytes_to_store & mask) + 1;
+        bytes_to_store_sizes.push(bytes_to_store);
+        encoded_bytes_to_store >>= 2;
+    }
+
+    // Because encoding and decoding are started on the same
+    // end of the byte, we need to reverse `bytes_to_store_sizes`
+    // to get sizes in order.
+    bytes_to_store_sizes.reverse();
+
+    bytes_to_store_sizes
+}
+
+impl<A, B, C> Storable for (A, B, C)
+where
+    A: Storable,
+    B: Storable,
+    C: Storable,
+{
+    // Touple will be serialized in the following form:
+    // <size_lengths (1B)> <size_a (1-4B)> <a_bytes> <size_b(1-4B)> <b_bytes> <c_bytes>
```

Comment:
If both `A` and `B` are fixed in size, then we don't need to store a size_lengths byte at all, right?

Reply:
Good point, the code is updated, PTAL.